internal/mongodoc/resource.go

```diff
@@ -0,0 +1,188 @@
+	// metadata.
+	Name string `bson:"name"`
+
+	// Revision identifies the specific revision of the resource.
+	Revision int `bson:"revision"`
+
+	// Fingerprint is the checksum of the resource file.
+	Fingerprint []byte `bson:"fingerprint"`
+
+	// Size is the size of the resource file, in bytes.
+	Size int64 `bson:"size"`
+
+	// BlobName holds the name that the resource blob is given in the
+	// blob store.
+	BlobName string
+
+	// UploadTime is the UTC timestamp from when the resource file was
+	// stored in the blob store.
+	UploadTime time.Time
+}
+
+// Validate ensures that the doc is valid.
+func (doc Resource) Validate() error {
+	if doc.CharmURL == nil {
+		return errgo.New("missing charm URL")
+	}
+	if doc.CharmURL.Revision != -1 {
+		return errgo.Newf("resolved charm URLs not supported (got revision %d)", doc.CharmURL.Revision)
+	}
+	if doc.CharmURL.Series != "" {
+		return errgo.Newf("series should not be set (got %q)", doc.CharmURL.Series)
+	}
+
+	if doc.Name == "" {
+		return errgo.New("missing name")
+	}
+	if doc.Revision < 0 {
+		return errgo.Newf("got negative revision %d", doc.Revision)
+	}
+
+	if len(doc.Fingerprint) == 0 {
+		return errgo.New("missing fingerprint")
+	}
+	fp, err := resource.NewFingerprint(doc.Fingerprint)
+	if err != nil {
+		return errgo.Notef(err, "bad fingerprint")
+	}
+	if err := fp.Validate(); err != nil {
+		return errgo.Notef(err, "bad fingerprint")
+	}
+
+	if doc.Size < 0 {
+		return errgo.Newf("got negative size %d", doc.Size)
+	}
+
+	if doc.BlobName == "" {
+		return errgo.New("missing blob name")
+	}
+
+	if doc.UploadTime.IsZero() {
+		return errgo.New("missing upload timestamp")
+	}
+
+	return nil
+}
+
+// HashString returns the resource's hex-encoded fingerprint.
+func (doc Resource) HashString() string {
+	return fmt.Sprintf("%x", doc.Fingerprint)
+}
+
+// SortResources sorts the provided resource docs.
+func SortResources(resources []*Resource) {
+	sort.Sort(resourcesByName(resources))
+}
+
+type resourcesByName []*Resource
+
+func (sorted resourcesByName) Len() int      { return len(sorted) }
+func (sorted resourcesByName) Swap(i, j int) { sorted[i], sorted[j] = sorted[j], sorted[i] }
+func (sorted resourcesByName) Less(i, j int) bool {
+	if sorted[i].Name < sorted[j].Name {
+		return true
+	}
+	if sorted[i].Name == sorted[j].Name && sorted[i].Revision < sorted[j].Revision {
+		return true
+	}
+	return false
+}
+
+// NewResourcesQuery formats the provided details into a mongo query
+// for the resource revisions associated with the identified *resolved*
+// charm URL (in the given channel).
+func NewResourcesQuery(cURL *charm.URL, channel params.Channel) bson.D {
+	query := bson.D{
+		{"resolved-charm-url", cURL},
+	}
+	if channel != params.NoChannel {
+		query = append(query, bson.DocElem{"channel", channel})
+	}
+	return query
+}
+
+// Resources identifies the set of resource revisions for a resolved
+// charm, relative to a specific channel,
+type Resources struct {
```

Comment:
I think I'd put this data in the base entity along with the other channel-related fields.
Something like this perhaps?

    // ChannelResources holds the set of resources currently
    // associated with each channel.
    ChannelResources map[params.Channel] []ResourceId

    type ResourceId struct {
        Name string
        Revision int
    }